src/java/org/apache/cassandra/index/sai/disk/v1/sortedterms/SortedTermsReader.java

```diff
@@ -112,123 +108,259 @@ public SortedTermsReader(@Nonnull FileHandle termsDataFileHandle,
-                                                                  "greater than the term count [%s]", pointId, meta.termCount));
-            long blockIndex = pointId >>> TERMS_DICT_BLOCK_SHIFT;
-            long blockAddress = blockOffsets.get(blockIndex);
-            termsInput.seek(blockAddress + termsDataFp);
-            this.pointId = (blockIndex << TERMS_DICT_BLOCK_SHIFT) - 1;
-            while (this.pointId < pointId && advance());
+            BytesRef skipTerm = readBytes(term);
+
+            currentBlockIndex = startingPointId >>> blockShift;
+            resetPosition();
+
+            if (compareTerms(currentTerm, skipTerm) == 0)
+                return startingPointId;
+
+            if (notInCurrentBlock(startingPointId, skipTerm))
+            {
+                long split = (endingPointId - startingPointId) >>> blockShift;
+                long splitPointId = startingPointId;
+                while (split > 0)
+                {
+                    currentBlockIndex = Math.min((splitPointId >>> blockShift) + split, blockOffsets.length() - 1);
+                    resetPosition();
+
+                    if (currentPointId >= endingPointId)
+                    {
+                        currentBlockIndex = (endingPointId - 1) >>> blockShift;
+                        resetPosition();
+                    }
+
+                    int cmp = compareTerms(currentTerm, skipTerm);
+
+                    if (cmp == 0)
+                        return currentPointId;
+
+                    if (cmp < 0)
+                        splitPointId = currentPointId;
+
+                    split /= 2;
+                }
+                // After we finish the binary search we need to move the block back till we hit a block that has
+                // a starting term that is less than or equals to the skip term
+                while (currentBlockIndex > 0 && compareTerms(currentTerm, skipTerm) > 0)
+                {
+                    currentBlockIndex--;
+                    resetPosition();
+                }
+            }
+
+            // Depending on where we are in the block we may need to move forwards to the starting point ID
+            while (currentPointId < startingPointId)
+            {
+                currentPointId++;
+                readTerm(currentPointId, currentTerm);
+                currentBlockIndex = currentPointId >>> blockShift;
+            }
+
+            // Move forward to the ending point ID, returning the point ID if we find our term
+            while (currentPointId < endingPointId)
+            {
+                if (compareTerms(currentTerm, skipTerm) >= 0)
+                    return currentPointId;
+                currentPointId++;
+                if (currentPointId == meta.termCount)
+                    return -1;
+                readTerm(currentPointId, currentTerm);
+                currentBlockIndex = currentPointId >>> blockShift;
```

Comment:
nit: We do this in a couple places...might be nice to have an `updateCurrentBlockIndex()` or an `advanceBlockIndex(pointId)` or something similar...

Reply:
`updateCurrentBlockIndex(long pointId)` works.